capstone_project/src/main/java/com/google/sps/data/PostAnalysis.java

```diff
@@ -0,0 +1,191 @@
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.cloud.language.v1.ClassificationCategory;
+import com.google.cloud.language.v1.ClassifyTextRequest;
+import com.google.cloud.language.v1.ClassifyTextResponse;
+import com.google.cloud.language.v1.Document;
+import com.google.cloud.language.v1.Document.Type;
+import com.google.cloud.language.v1.LanguageServiceClient;
+import com.google.cloud.language.v1.Sentiment;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.StringTokenizer;
+import javax.servlet.http.HttpServletRequest;
+
+/**
+ * Analyzes user posts to determine category classifications, which are used to provide users with
+ * resources in response to the specific categories
+ */
+public class PostAnalysis {
+  private DatastoreService datastore;
+  private LanguageServiceClient languageService;
+  private double sentimentScore;
+  private final List<String> categories;
+  private final List<String> resources;
+  public final int MIN_TOKENS = 20;
+
+  public static class Builder {
+    private LanguageServiceClient languageService;
+
+    public static Builder builder() {
+      return new Builder();
+    }
+
+    public Builder languageService(LanguageServiceClient languageService) {
+      this.languageService = languageService;
+      return this;
+    }
+
+    public PostAnalysis build() throws IOException {
+      if (this.languageService == null) {
+        this.languageService(LanguageServiceClient.create());
+      }
+      return new PostAnalysis(this);
+    }
+  }
+
+  private PostAnalysis(Builder builder) throws IOException {
+    datastore = DatastoreServiceFactory.getDatastoreService();
+    languageService = builder.languageService;
+    categories = new ArrayList<>();
+    resources = new ArrayList<>();
+  }
+
+  /**
+   * Returns sentiment score calculated for the message
+   *
+   * @return sentiment score
+   */
+  public double getSentimentScore() {
+    return sentimentScore;
+  }
+
+  /**
+   * Returns categories that the message falls under
+   *
+   * @return list of categories
+   */
+  public List<String> getCategories() {
+    return Collections.unmodifiableList(categories);
+  }
+
+  /**
+   * Returns the resources suggested for the user, based on the message's categories
+   *
+   * @return resources suggested
+   */
+  public List<String> getResources() {
+    return Collections.unmodifiableList(resources);
+  }
+
+  /** Set LanguageServiceClient from Google Natural Language API */
+  public void setLanguageServiceClient(LanguageServiceClient languageService) throws IOException {
```

Comment:
why would this throw an exception?

Reply:
I have removed this method as I set the LanguageServiceClient solely through the Builder class